feat(ngModel): provide ng-empty and ng-not-empty CSS classes

petebacondarwin opened the pull request:
I believe that this feature should be enough to fix #10050

petebacondarwin committed:
test(ngModel): remove jankiness test

It is no longer appropriate to test this here as $animate takes care of
queueing up CSS class changes into a single update.

gkalpak commented on test/ng/directive/ngModelSpec.js:
```diff
@@ -1182,46 +1182,6 @@ describe('ngModel', function() {
-        expect(removeClassCallCount).toBe(0);
-
-        dealoc(element);
-      }));
```
Out of curiosity, why was this removed ?

gkalpak commented:
I wonder if people will ask for this to be propagated to a parent `FormController`...

petebacondarwin commented:
I was trying not to mention that :-P

Narretz commented:
For the original use case, you would then use`ng-pristine` and `ng-not-empty` together?

Narretz commented on src/ng/directive/ngModel.js:
```diff
@@ -316,6 +318,17 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
   };
 
+  this.$$updateEmptyClasses = function(value) {
+    if (this.$isEmpty(value)) {
```
Should `this` be `ctrl`?

petebacondarwin replied:
It doesn't really matter because we have control over when it is called but I agree it is probably better for consistency.

Narretz commented on test/ng/directive/ngModelSpec.js:
```diff
@@ -1368,6 +1328,26 @@ describe('ngModel', function() {
   describe('CSS classes', function() {
     var EMAIL_REGEXP = /^[a-z0-9!#$%&'*+\/=?^_`{|}~.-]+@[a-z0-9]([a-z0-9-]*[a-z0-9])?(\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)*$/i;
 
+    it('should set ng-empty or ng-not-empty the view value changes', inject(function($compile, $rootScope, $sniffer) {
```
.. ng-not-empty (when) the view value changes ..

Narretz replied:
Also, line lengths? I never know what the limit should be. 100?

gkalpak replied:
Theoretically, yes, they should be 100 chars max :smiley:
See: [CONTRIBUTING.md#-coding-rules][1]

[1]: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#-coding-rules

petebacondarwin replied:
Our jshint rules are now set to 200 for line length :-)

gkalpak replied:
Pure evil :-P

petebacondarwin committed:
feat(ngModel): provide ng-empty and ng-not-empty CSS classes

If the `$viewValue` is empty then the `ng-empty` CSS class is applied
to the input. Conversely, if it is not empty the `ng-not-empty` CSS class
is applied. Emptiness is ascertained by calling `NgModelController.$isEmpty()`

Closes #10050
Closes #12848

petebacondarwin commented:
Thanks for the reviews @Narretz and @gkalpak. I added a commit.

petebacondarwin commented:
@Narretz  and @gkalpak can you take another look?

Narretz commented on src/ng/directive/ngModel.js:
```diff
@@ -316,6 +318,17 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
     return isUndefined(value) || value === '' || value === null || value !== value;
   };
 
+  this.$$updateEmptyClasses = function(value) {
```
This could be simplified by always using ctrl.$viewValue instead of passing it as an arg. It should only be the viewValue anyway. You'll have to change the order on line https://github.com/angular/angular.js/pull/12848/files#diff-ae1963e24d6328c3aa019c0c1b9f4432R851 though, because otherwise it'll use the old viewValue

petebacondarwin replied:
Actually it would have to use `$$lastCommittedViewValue` and we would have to change the order as you suggest. I am not that keen on doing this though as I don't feel that not passing in an argument makes it that much simpler.

Narretz commented:
Regarding my question
> For the original use case, you would then use ng-pristine and ng-not-empty together?

because only ng-empty doesn't work if I understood the issue correctly.

petebacondarwin commented:
@Narretz: Yes, you would use a combination of the CSS classes for the original use case, but this seemed reasonable and intuitive to me.

petebacondarwin commented:
Thanks for the review @Narretz.
So assuming that we don't simplify the method and that the original use case is covered by a combination of CSS classes. Is this PR ready to merge?

Narretz commented:
Yep, feel free to merge!
Am 21.09.2015 22:45 schrieb "Pete Bacon Darwin" <notifications@github.com>:

> Thanks for the review @Narretz <https://github.com/Narretz>.
> So assuming that we don't simplify the method and that the original use
> case is covered by a combination of CSS classes. Is this PR ready to merge?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular.js/pull/12848#issuecomment-142101707>.
>
